README.md
spacetime
=========

R classes and methods for spatio-temporal data, see <http://cran.r-project.org/web/packages/spacetime/index.html> and <http://www.jstatsoft.org/v51/i07/>


inst/ChangeLog
commit 3083307fc8ebf938673028cf7ffc3ccff6a4cac6
Author: Edzer Pebesma <edzer.pebesma@uni-muenster.de>
Date:   Fri Apr 24 13:56:40 2015 +0200

    upgrade to match sp 1.1-0

commit 5e7950336984a567c0f07abb019930dbaf068abb
Author: Edzer Pebesma <edzer.pebesma@uni-muenster.de>
Date:   Tue Mar 17 14:21:20 2015 +0100

    docs for expand.range

commit 58d8db3d8863eee96aeea1cf458fbee2e42b32db
Author: Edzer Pebesma <edzer.pebesma@uni-muenster.de>
Date:   Tue Mar 17 14:18:44 2015 +0100

    range of at was too small; report by Roland Harhoff

commit af552c7e902b69979f309bcf1f795bc7cf782f11
Author: Edzer Pebesma <edzer.pebesma@uni-muenster.de>
Date:   Tue Mar 17 13:56:53 2015 +0100

    add drop=FALSE to data.frame selector

commit 8ac495755ea13ef90b9e378ec26b963ea316174b
Author: Edzer Pebesma <edzer.pebesma@uni-muenster.de>
Date:   Fri Mar 13 16:20:47 2015 +0100

    adding +ellps= to CRS arguments, for PROJ 4.9.1

commit 393bb24344b7b3648086ecadf5bf287bdea4fcde
Author: Edzer Pebesma <edzer.pebesma@uni-muenster.de>
Date:   Thu Mar 5 00:12:50 2015 +0100

    replace unstack with reshape; clean check

commit 45bf1588eef7e2ddbf07f843333d72f0a99d0e4f
Author: Edzer Pebesma <edzer.pebesma@uni-muenster.de>
Date:   Wed Mar 4 20:27:51 2015 +0100

    stplot mode = "xt" for factor: plots as number but no longer crashes

commit b7e51d1720dace123a73669ee04b77255a0b9c09
Author: Edzer Pebesma <edzer.pebesma@uni-muenster.de>
Date:   Wed Mar 4 19:07:27 2015 +0100

    stplot factor variables when mode = "xy"

commit 24671c67b17293749937d7c955a80ce8e12ac619
Author: Edzer Pebesma <edzer.pebesma@uni-muenster.de>
Date:   Wed Feb 11 10:51:18 2015 +0100

    make stsdf[1,1] do sth more sensible

commit e7d26c403722f6dfe08f27bbbe82644e2f48976e
Author: Edzer Pebesma <edzer.pebesma@uni-muenster.de>
Date:   Wed Jan 28 12:13:49 2015 +0100

    improve error message

commit f37ad515f35e21f9351c290437f2948f34b5f894
Author: Edzer Pebesma <edzer.pebesma@uni-muenster.de>
Date:   Wed Jan 28 10:25:09 2015 +0100

    enable check on duplicate attributes for STSDF

commit 995dd6d742143bf9fcc4d43c19c358f9a4bc626c
Author: Edzer Pebesma <edzer.pebesma@uni-muenster.de>
Date:   Tue Jan 27 23:34:52 2015 +0100

    adopt new behaviour for STFDF too

commit 08fcfb9f1555e283c82107256c055a5bead0b265
Author: Edzer Pebesma <edzer.pebesma@uni-muenster.de>
Date:   Tue Jan 27 23:18:42 2015 +0100

    add timeIndex to xts output on single pt subset; report from Jon Skoien;

commit 68d9420d6a9bad79f3b84c656cbfc9f670fe5a28
Author: Edzer Pebesma <edzer.pebesma@uni-muenster.de>
Date:   Sat Jan 24 10:19:17 2015 +0100

    remove rgeos dependency in sto.Rnw, add DE to air.rda

commit 097d3a0fbc63f2ef7773825c07d35eafa3ab3cba
Author: Edzer Pebesma <edzer.pebesma@uni-muenster.de>
Date:   Wed Jan 21 23:20:34 2015 +0100

    make rgeos requirement conditional

commit cddc6c9ac45dfc7664c7e5b3adbcc65b0f499961
Author: Edzer Pebesma <edzer.pebesma@uni-muenster.de>
Date:   Wed Jan 21 09:08:18 2015 +0100

    bump version

commit 218a99f5b7b475353f8b37246c945c47f881f6aa
Author: Edzer Pebesma <edzer.pebesma@uni-muenster.de>
Date:   Mon Jan 19 14:55:57 2015 +0100

    correct timeIndex order for as.data.frame.STS, reported by Jon Skoien

commit 6502ebfd51b6704669c46d26c8dd2b7304c630ef
Author: Edzer Pebesma <edzer.pebesma@uni-muenster.de>
Date:   Sat Jan 17 23:05:15 2015 +0100

    conditional require(rgeos) in example

commit 3f4f8a5c3e6048393ad14fd4444ec03773207dc9
Author: Edzer Pebesma <edzer.pebesma@uni-muenster.de>
Date:   Thu Jan 15 13:05:22 2015 +0100

    more spTransform

commit a29c98f3c80a784967bc319b5c63a9a725b631cd
Author: Edzer Pebesma <edzer.pebesma@uni-muenster.de>
Date:   Thu Jan 15 13:03:49 2015 +0100

    title case; address rgdal requirement for spTransform

commit e6073767b1e303dcb20c99cfe635d762bdd36eda
Author: Edzer Pebesma <edzer.pebesma@uni-muenster.de>
Date:   Thu Jan 15 11:51:11 2015 +0100

    update test output

commit 5b36a6916bc8af4cd696bb7742b0e0f03abe5b25
Author: Edzer Pebesma <edzer.pebesma@uni-muenster.de>
Date:   Wed Jan 14 17:33:25 2015 +0100

    typo...

commit f74c045284b795bc0e4ad1e167608dfa74a21674
Author: Edzer Pebesma <edzer.pebesma@uni-muenster.de>
Date:   Wed Jan 14 16:59:45 2015 +0100

    update test output

commit 0604dd381b7e2e10f97bbc3d5b26c8119c744a8f
Author: Edzer Pebesma <edzer.pebesma@uni-muenster.de>
Date:   Wed Jan 14 16:41:54 2015 +0100

    prevent @sp attributes getting dropped (lost)

commit a144894fbb8590a701cec4f99ae1f6f09d896916
Author: Edzer Pebesma <edzer.pebesma@uni-muenster.de>
Date:   Sun Dec 14 13:18:17 2014 +0100

    correct aggregate with args STIDF, STF; add time chk

commit 4c8cd4f1d6220dc43634e4333e18393a7d8d87e9
Author: Edzer Pebesma <edzer.pebesma@uni-muenster.de>
Date:   Sat Nov 22 12:26:03 2014 +0100

    correct bug in stConstruct() when STIDF has non-chronological records

commit 1969b0c7ba6197d96d4541a15a71a84a19291803
Author: Edzer Pebesma <edzer.pebesma@uni-muenster.de>
Date:   Thu Nov 20 11:26:02 2014 +0100

    vignette pdf

commit 850c3068348815205937350d9dad048ec3e18cfd
Author: Edzer Pebesma <edzer.pebesma@uni-muenster.de>
Date:   Thu Nov 20 11:22:36 2014 +0100

    update date

commit 2076a4e3b24c5f2aa8e029eca1a28f9d05f39d52
Author: Edzer Pebesma <edzer.pebesma@uni-muenster.de>
Date:   Thu Nov 20 11:20:35 2014 +0100

    aggregate() for STI and STF with polygons did not work

commit d903c16d831c43ab777a552b7c31687e5046396f
Author: Edzer Pebesma <edzer.pebesma@uni-muenster.de>
Date:   Thu Oct 30 10:55:32 2014 +0100

    move rgeos to Suggests; improve wind plots in vignette

commit da80be7b754e3b613107845166687b02deeb96dc
Author: Edzer Pebesma <edzer.pebesma@uni-muenster.de>
Date:   Mon Oct 27 19:26:35 2014 +0100

    update the time plots

commit c1f9a7480241deeb177e345332e4045236e9c02e
Author: Edzer Pebesma <edzer.pebesma@uni-muenster.de>
Date:   Mon Oct 27 19:17:24 2014 +0100

    change argument to returnEOFs

commit 1783943b6aa395ee0296bed53e88126b51909fd3

demo/edc.R
# exercises for ifgi - Esri development center workshop
# held Muenster, Mar 17-18 2011

###################################################
### chunk number 26: 
###################################################
#line 577 "spacetime.Rnw"

# here is an example where data in long format that come
# as a table in package plm, are combined (through USA state names)
# with state boundaries in package maps:

# first obtain the state boundaries, and their IDs (state names):
library(maps)
states.m = map('state', plot=FALSE, fill=TRUE)
IDs <- sapply(strsplit(states.m$names, ":"), function(x) x[1])
IDs
    
# then, convert the states to a SpatialPolygons object:
library(maptools)
states = map2SpatialPolygons(states.m, IDs=IDs)
plot(states)

# next, import the Produc data, a table with space and time in
# columns 1 and 2:
library(plm)
data(Produc)
summary(Produc)

# then, construct the STFDF object:
library(spacetime)
yrs = 1970:1986
time = xts(1:17, as.POSIXct(paste(yrs, "-01-01", sep=""), tz = "GMT"))
# deselect District of Columbia, polygon 8, which is not present in Produc:
Produc.st = STFDF(states[-8], time, Produc[(order(Produc[2], Produc[1])),])

# and plot it:
stplot(Produc.st[,,"unemp"], yrs)


###################################################
### chunk number 28: 
###################################################
#line 638 "spacetime.Rnw"

# spatio-temporal interpolation: the Irish wind data set
# first: import the data:
library(gstat)
data(wind)
# this imports two data.frame's: wind.loc and wind:
summary(wind.loc)
summary(wind)
wind.loc$y = as.numeric(char2dms(as.character(wind.loc[["Latitude"]])))
wind.loc$x = as.numeric(char2dms(as.character(wind.loc[["Longitude"]])))
coordinates(wind.loc) = ~x+y
proj4string(wind.loc) = "+proj=longlat +datum=WGS84"
# now wind.loc is converted into a georeferenced SpatialPointsDataFrame object:
summary(wind.loc)


###################################################
### chunk number 29: 
###################################################
#line 649 "spacetime.Rnw"

# load a map reference for these data, and plot the wind stations:
library(mapdata)
plot(wind.loc, xlim = c(-11,-5.4), ylim = c(51,55.5), axes=T, col="red",
	cex.axis =.7)
map("worldHires", add=T, col = grey(.5))
text(coordinates(wind.loc), pos=1, label=wind.loc$Station, cex=.7)


###################################################
### chunk number 30: 
###################################################
#line 666 "spacetime.Rnw"

# check the wind data: they are in space-wide form: stations form columns
wind[1:3,]


###################################################
### chunk number 31: 
###################################################
#line 671 "spacetime.Rnw"

# let us reference the time information into a time class:
wind$time = ISOdate(wind$year+1900, wind$month, wind$day)
summary(wind$time)

# the next few commands are some statistical magic; wind is sqrt-transformed
# and a smooth pattern is subtracted from the raw numbers:
wind$jday = as.numeric(format(wind$time, '%j'))
stations = 4:15
windsqrt = sqrt(0.5148 * wind[stations]) # knots -> m/s
Jday = 1:366
daymeans = apply(sapply(split(windsqrt - mean(windsqrt), wind$jday), mean), 2,  mean)
meanwind = lowess(daymeans ~ Jday, f = 0.1)$y[wind$jday]
velocities = apply(windsqrt, 2, function(x) { x - meanwind })

###################################################
### chunk number 32: 
###################################################
#line 685 "spacetime.Rnw"

# order locations to order of columns in wind;
# connect station names to location coordinates
wind.loc = wind.loc[match(names(wind[4:15]), wind.loc$Code),]
pts = coordinates(wind.loc[match(names(wind[4:15]), wind.loc$Code),])
rownames(pts) = wind.loc$Station
pts = SpatialPoints(pts)

# convert to utm zone 29, to be able to do interpolation in
# proper Euclidian (projected) space:
proj4string(pts) = "+proj=longlat +datum=WGS84"
library(rgdal)
utm29 = CRS("+proj=utm +zone=29 +datum=WGS84")
t = xts(1:nrow(wind), wind$time)
pts = spTransform(pts, utm29)
# create the spacetime object:
# note the t() in:
w = STFDF(pts, t, data.frame(values = as.vector(t(velocities))))

# import, convert, and project the backdrop (country boundaries) data too:
library(maptools)
m = map2SpatialLines(
	map("worldHires", xlim = c(-11,-5.4), ylim = c(51,55.5), plot=F))
proj4string(m) = "+proj=longlat +datum=WGS84"
m = spTransform(m, utm29)

# setup a grid covering the region of m:
grd = SpatialPixels(SpatialPoints(makegrid(m, n = 300)),
	proj4string = proj4string(m))
# select data (arbitrarily) for april 1961:
w = w[, "1961-04"]
# 10 prediction time points, evenly spread over this month:
n = 10
tgrd = xts(1:n, seq(min(index(w)), max(index(w)), length=n))

# using a separable exponential covariance model, 
# with ranges 750 km and 1.5 day, do a space-time interpolation:
v = list(space = vgm(0.6, "Exp", 750000), time = vgm(1, "Exp", 1.5 * 3600 * 24))
pred = krigeST(sqrt(values)~1, w, STF(grd, tgrd), v)
wind.ST = STFDF(grd, tgrd, data.frame(sqrt_speed = pred))

###################################################
### chunk number 33: 
###################################################
#line 727 "spacetime.Rnw"

# plot the results:
layout = list(list("sp.lines", m, col='grey'),
	list("sp.points", pts, first=F, cex=.5))
stplot(wind.ST, col.regions=bpy.colors(),
	par.strip.text = list(cex=.5), sp.layout = layout)

# Exercises:
# 1. repeat the procedure but interpolate the daily 12:00 values in May 1962
# 2. plot, from the results obtained, only the first three days
# 3. repeat the procedure, but use a grid with 1000 grid cells instead of 300.
